src/units/abilities.cpp

```diff
@@ -885,33 +879,118 @@ std::string attack_type::weapon_specials(bool only_active, bool is_backstab) con
 	}
 	std::string weapon_abilities;
 	std::set<std::string> checking_name;
-	assert(display::get_singleton());
-	const unit_map& units = display::get_singleton()->get_units();
 	if(self_){
-		for (const config::any_child sp : self_->abilities().all_children_range()){
-			const bool active = check_self_abilities_impl(shared_from_this(), other_attack_, sp.cfg, self_, self_loc_, AFFECT_SELF, sp.key);
-			add_name(weapon_abilities, active, sp, checking_name, false);
+		weapon_specials_impl_self(weapon_abilities, self_, shared_from_this(), other_attack_, self_loc_, AFFECT_SELF, checking_name);
+		weapon_specials_impl_adj(weapon_abilities, self_, shared_from_this(), other_attack_, self_loc_, AFFECT_SELF, checking_name);
+	}
+	if(other_){
+		weapon_specials_impl_self(weapon_abilities, other_, other_attack_, shared_from_this(), other_loc_, AFFECT_OTHER, checking_name);
+		weapon_specials_impl_adj(weapon_abilities, other_, other_attack_, shared_from_this(), other_loc_, AFFECT_OTHER, checking_name);
+	}
+	if(!weapon_abilities.empty() && !res.empty()) {
+		weapon_abilities = ", \n" + weapon_abilities;
+		res += weapon_abilities;
+	} else if (!weapon_abilities.empty()){
+		res = weapon_abilities;
+	}
+	return res;
+}
+
+static void add_string(std::string& weapon_abilities, std::string& self_abilities, const std::string temp_abilities, std::set<std::string>& checking_name)
+{
+	weapon_abilities = "";
+	checking_name.clear();
```

Comment:
What do weapon_abilities and checking_name have to do with adding a string? This seems to be part of the flow logic of the caller instead.

Reply:
i must rename that clear_temp_string instead.

Reply:
Irrespective of the names, what do these two variables have to do with this function; why are they involved in it? They're not inputs, and they don't really feel like outputs either - they both get changed, but their final value doesn't depend on any of the inputs to the function.

Reply:
Reviewing ea7bcc0c452, I still don't understand why checking_name is involved in this function.

Reply:
This is for improbable, but not impossible, cases where weapons from different sources (an ability carried by self, and another by teacher for example), would have the same name, if in the window, the name should only be mentioned only once, in the tooltip, it seems relevant to point out that the stats are influenced by two weapons possessed and taught, the example of a charge type ability can also illustrate this when two units fight with the same capacity of this type (active both in attack and defense), the player sees the attack stats of his unit influenced by his own ability but also that of his opponent. To allow this, checking_name must be cleared after each block (self, teacher or opponent) is written.

Reply:
I think I understand, but feel it could be refactored to be simpler. If I understand the current implementation correctly:
* When `add_name_list` returns, `checking_name` is always empty. Although the `clear()` is in a conditional block, that condition will always be true except for the times that `checking_name` is already empty when `add_name_list` is entered.
* In `weapon_specials_value` the only times that `checking_name` could be non-empty on entry to a `weapon_specials_impl_*` function are the ones for the `"Self: "` and `"Opponent: "` blocks.
* There are more places that it could be non-empty in `weapon_specials`.
* In `weapon_specials_value` there's only one time that the value of `checking_name` returned from a `weapon_specials_impl_*` function is used, for the `"Opponent: "` block.

Reply:
for the first point, conditionnal block not ever true, if weapon_specials_impl_* return a empty temp_string, and preffix not added to non-existent list.
. Correct "Self: " has weapon specials and abilities like weapon owned by unit in list, and opponent has all weapon special or abilities like weapon applied used by opponent(owned or teached) and applied to unit

. Correct weapon_specials apllied to windows and origin to each weapon\abilities) non specifed, a name must be addedone time only.

. Yes because opponent is the only case or two weapon_specials_impl_* are used.

Reply:
I'm wondering if we can omit special handling for this edge-case. For the `"Self: "` case, I think it only happens if the unit has an ability called `x` and a weapon special also called `x` - in which case, showing `x, x` will look odd but possibly be easier for the player to understand than combining them into `x` when two badly-named things are active.

For the `"Other: "` string, please would you give an example of an ability that should be returned by `weapon_specials_impl_adj`? I'm wondering if those will already be in "Enemy's Teachers: ".

Reply:
using "Self: " for abilities only, ok. 

For "Opponent: " a leadership charge teached to opponent but appied also to attacker must appear in in tooltip, and the "opponent: " string is here for said what damage *2by abilities used by opponent, same if teached.